lib/Sema/DerivedConformanceComparable.cpp

```diff
@@ -0,0 +1,411 @@
+  auto bParam = args->get(1);
+
+  auto enumDecl = cast<EnumDecl>(aParam->getType()->getAnyNominal());
+
+  // Generate the conversion from the enums to integer indices.
+  SmallVector<ASTNode, 6> statements;
+  DeclRefExpr *aIndex = convertEnumToIndex(statements, parentDC, enumDecl,
+                                           aParam, ltDecl, "index_a");
+  DeclRefExpr *bIndex = convertEnumToIndex(statements, parentDC, enumDecl,
+                                           bParam, ltDecl, "index_b");
+
+  // Generate the compare of the indices.
+  FuncDecl *cmpFunc = C.getLessThanIntDecl();
+  assert(cmpFunc && "should have a < for int as we already checked for it");
+
+  auto fnType = cmpFunc->getInterfaceType()->castTo<FunctionType>();
+
+  Expr *cmpFuncExpr;
+  if (cmpFunc->getDeclContext()->isTypeContext()) {
+    auto contextTy = cmpFunc->getDeclContext()->getSelfInterfaceType();
+    Expr *base = TypeExpr::createImplicitHack(SourceLoc(), contextTy, C);
+    Expr *ref = new (C) DeclRefExpr(cmpFunc, DeclNameLoc(), /*Implicit*/ true,
+                                    AccessSemantics::Ordinary, fnType);
+
+    fnType = fnType->getResult()->castTo<FunctionType>();
+    cmpFuncExpr = new (C) DotSyntaxCallExpr(ref, SourceLoc(), base, fnType);
+    cmpFuncExpr->setImplicit();
+  } else {
+    cmpFuncExpr = new (C) DeclRefExpr(cmpFunc, DeclNameLoc(),
+                                      /*implicit*/ true,
+                                      AccessSemantics::Ordinary,
+                                      fnType);
+  }
+
+  TupleTypeElt abTupleElts[2] = { aIndex->getType(), bIndex->getType() };
+  TupleExpr *abTuple = TupleExpr::create(C, SourceLoc(), { aIndex, bIndex },
+                                         { }, { }, SourceLoc(),
+                                         /*HasTrailingClosure*/ false,
+                                         /*Implicit*/ true,
+                                         TupleType::get(abTupleElts, C));
+
+  auto *cmpExpr = new (C) BinaryExpr(
+      cmpFuncExpr, abTuple, /*implicit*/ true,
+      fnType->castTo<FunctionType>()->getResult());
+  statements.push_back(new (C) ReturnStmt(SourceLoc(), cmpExpr));
+
+  BraceStmt *body = BraceStmt::create(C, SourceLoc(), statements, SourceLoc());
+  return { body, /*isTypeChecked=*/true };
+}
+
+/// Derive the body for an '==' operator for an enum where at least one of the
+/// cases has associated values.
+static std::pair<BraceStmt *, bool>
+deriveBodyComparable_enum_hasAssociatedValues_lt(AbstractFunctionDecl *ltDecl, void *) {
+  auto parentDC = ltDecl->getDeclContext();
+  ASTContext &C = parentDC->getASTContext();
+
+  auto args = ltDecl->getParameters();
+  auto aParam = args->get(0);
+  auto bParam = args->get(1);
+
+  Type enumType = aParam->getType();
+  auto enumDecl = cast<EnumDecl>(aParam->getType()->getAnyNominal());
+
+  SmallVector<ASTNode, 6> statements;
+  SmallVector<ASTNode, 4> cases;
+  unsigned elementCount = 0;
```

Comment:
Let's use `enumDecl->getAllElements().size()` in the comparison below instead of this variable.

Reply:
`enumDecl->getAllElements()` returns a generator, which does not support a `size()` operation. i couldn’t find any differenceing API in `STLExtras.h` either